src/firmware/app/control/trajectory_planner.c

```diff
@@ -0,0 +1,330 @@
+    unsigned int num_segments, float* velocity_profile,
+    float* max_allowable_speed_profile, const float arc_segment_length,
+    const float max_allowable_acceleration, const float max_allowable_speed)
+{
+    for (unsigned int i = 1; i < num_segments; i++)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const float temp_vel =
+            (float)sqrt(pow(velocity_profile[i - 1], 2) +
+                        2 * arc_segment_length * max_allowable_acceleration);
+
+        // If the new speed is greater than the max allowable, reduce it to the max
+        if (temp_vel > max_allowable_speed_profile[i])
+        {
+            velocity_profile[i] = max_allowable_speed_profile[i];
+        }
+        // If the new speed is greater than the maximum limit of the robot, reduce it
+        // to the limit
+        else if (temp_vel >= max_allowable_speed)
+        {
+            velocity_profile[i] = max_allowable_speed;
+        }
+        else
+        {
+            velocity_profile[i] = temp_vel;
+        }
+    }
+}
+
+void app_trajectory_planner_generate_backwards_continuous_velocity_profile(
+    unsigned int num_segments, float* forwards_continuous_velocity_profile,
+    const float arc_segment_length, const float max_allowable_acceleration)
+{
+    for (unsigned int i = num_segments - 1; i > 0; i--)
+    {
+        // Vf = sqrt( Vi^2 + 2*constant_segment_length*max_acceleration)
+        const float temp_vel =
+            (float)sqrt(pow(forwards_continuous_velocity_profile[i], 2) +
+                        2 * arc_segment_length * max_allowable_acceleration);
+
+        if (forwards_continuous_velocity_profile[i - 1] > temp_vel)
+        {
+            forwards_continuous_velocity_profile[i - 1] = temp_vel;
+        }
+    }
+}
+
+void app_trajectory_planner_generate_time_profile(TrajectoryElement_t* traj_elements,
+                                                  const float num_segments,
+                                                  const float arc_segment_length,
+                                                  float* velocity_profile)
+{
+    traj_elements[0].time = 0.0;
+    for (unsigned int i = 0; i < num_segments - 1; i++)
+    {
+        traj_elements[i + 1].time =
+            traj_elements[i].time +
+            (2 * arc_segment_length) / (velocity_profile[i] + velocity_profile[i + 1]);
+    }
+}
+
+void app_trajectory_planner_reverse_trajectory_direction(TrajectoryElement_t* forwards,
+                                                         unsigned int num_segments)
+{
+    TrajectoryElement_t reverse[TRAJECTORY_PLANNER_MAX_NUM_ELEMENTS];
+    const float path_duration = forwards[num_segments - 1].time;
+
+    for (unsigned int i = 0; i < num_segments; i++)
+    {
+        // Reverse the positions
+        reverse[(num_segments - 1) - i].position = forwards[i].position;
+        reverse[(num_segments - 1) - i].time     = fabs(path_duration - forwards[i].time);
```

Comment:
Why fabs? Why isn't path_duration > forwards[i].time always true?

Reply:
Turns out I didn't need it! Changed :+1: